Adds support for Bitbucket Cloud Code Insights

marvin-w opened the pull request:
### Intention

This PR intends to provide support for the code insights feature for bitbucket cloud.

### History

Since about 2 months bitbucket cloud also has a code insights feature that one can use. After checking the differences between the cloud and the server implementation it is however not possible to completely reuse the server logic due to renamed/missing fields in the cloud version.

I haven't changed any public facing API (configs, properties, decorators) in this PR which allows @mc1arke to keep it up to date with the other sonarqube versions.

### Additions

* The API is fully functional now for both the server and the cloud version. One development limitation is that Bitbucket actually checks the URL and doesn't allow localhost URLs. That's why someone testing this with a local instance or an instance whose DNS isn't publicly accessible (not confirmed) might have troubles.
* I tried to create a wiki page for this new decorator, however according to https://github.community/t/how-to-fork-a-wiki-and-send-a-pr/2006 it is currently not possible to fork a wiki repository. Is it possible to be granted access to feature branches for the wiki git - then I could create a PR. Alternatively, setting up a dedicated wiki repository could also make sense in case other people are interested in contributing to it.

### Screenshots

![image](https://user-images.githubusercontent.com/708887/83976314-99dde280-a8f9-11ea-848e-393860c02a0b.png)

![image](https://user-images.githubusercontent.com/708887/83976327-a5c9a480-a8f9-11ea-8d69-e068e9f08f50.png)

marvin-w commented:
@mc1arke I'm done with this. Feel free to review it once you have some time.

@singhsoldier13, @goober, @jburgel-davinci, @yeroc If there is a chance it would be awesome if you could also go for an additional test run with this version of bitbucket cloud PR decoration.

For the token: Please use 

`base64(username:password)`

For the URL please use `https://api.bitbucket.org`

marvin-w commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketServerClient.java:
```diff
@@ -122,49 +108,15 @@ public boolean supportsCodeInsights() {
-        return configuration.get(BitbucketServerPullRequestDecorator.PULL_REQUEST_BITBUCKET_TOKEN)
-                .orElseThrow(() -> new IllegalArgumentException("Personal Access Token for Bitbucket Server is missing"));
+        return new OkHttpClient.Builder()
+                .addInterceptor(chain -> {
```
@mc1arke The code before used the authenticator call. However, the header for the authorization was never passed to the actual requests thus I changed it to interceptor based.

mc1arke replied:
👍

marvin-w commented on src/main/java/com/github/mc1arke/sonarqube/plugin/CommunityBranchPlugin.java:
```diff
@@ -147,10 +147,6 @@ public void load(CoreExtension.Context context) {
                             "Repository Slug see for example https://docs.atlassian.com/bitbucket-server/rest/latest/bitbucket-rest.html")
                             .type(PropertyType.STRING).build(),
 
-                    PropertyDefinition.builder(BitbucketServerPullRequestDecorator.PULL_REQUEST_BITBUCKET_USER_SLUG).category(PULL_REQUEST_CATEGORY_LABEL).subCategory(BITBUCKET_INTEGRATION_SUBCATEGORY_LABEL)
```
I deleted this as we don't use it anymore and it just confuses the users

mc1arke replied:
👍

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/AbstractBitbucketClient.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    public void setConfiguration(UnifyConfiguration configuration) {
+        this.configuration = configuration;
```
This is likely to cause an issue: the instances created by the Pico container are singletons so you're risking a race condition by modifying the internal state of a decorator whilst it's potentially in use in another thread. You'll need to ensure your approach is stateless by either converting this to a generator/factory pattern, or using injected instances that are state aware (e.g. creating a `UnifyConfiguration` when required based on using an injected `Configuration` and `ScannerContext` on-demand)

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/AbstractBitbucketClient.java:
```diff
@@ -0,0 +1,62 @@
+
+import java.io.IOException;
+
+abstract class AbstractBitbucketClient {
```
There are no `abstract` methods, so this class doesn't need to be `abstract`. Based on my review comment on the `setConfiguration` method in this class, you may actually want to make an abstract `createConfiguration` method that the relevant extending classes then implement.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/AbstractBitbucketClient.java:
```diff
@@ -0,0 +1,62 @@
+abstract class AbstractBitbucketClient {
+    private UnifyConfiguration configuration;
+
+    ObjectMapper getObjectMapper() {
```
Since this doesn't return the same instance on multiple calls, it shouldn't be a `get` method, instead being something like `create`

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketCloudClient.java:
```diff
@@ -0,0 +1,110 @@
+            if (response.body() != null) {
+                error = response.body().string();
+            }
+            throw new RuntimeException("Bitbucket returned the following error: " + error);
```
This provides a poor error message if the response body is `null`. Could we clarify it to say that no response body has been returned if the `error` is `null`

mc1arke replied:
Any chance of using something slightly less generic than a `RuntimeException`. An `IllegalStateException` may suit if you don't have anything more specific.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/model/IAnnotation.java:
```diff
@@ -0,0 +1,25 @@
+/**
+ * Interface for reusing models between the cloud and the server version
+ */
+public interface IAnnotation {
```
I'm not a fan of prefixing interfaces with `I`. I know there's a class called `Annotation`, but is there any chance we can rename it, or give this interface another name that isn't based on the type of class?

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/model/cloud/CloudCreateReportRequest.java:
```diff
@@ -0,0 +1,116 @@
+    @JsonProperty("result")
+    private final String result;
+
+    @JsonProperty("remote_link_enabled")
```
Any reason for the whitespace above this but not other fields?

mc1arke commented on src/test/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketClientFacadeUnitTest.java:
```diff
@@ -0,0 +1,201 @@
+import java.time.Instant;
+import java.util.ArrayList;
+
+import static org.junit.Assert.*;
```
Please don't use star imports

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/model/cloud/CloudCreateReportRequest.java:
```diff
@@ -0,0 +1,116 @@
+    private final String result;
+
+    @JsonProperty("remote_link_enabled")
+    private final Boolean remoteLinkEnabled = true;
```
Any reason for this not being a primitive `boolean`?

marvin-w commented:
@mc1arke I did a major refactoring considering your point about the race condition. Would you mind having a quick check again - I'd then finish all the tests and take care of sonar.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -74,10 +73,10 @@
                     .collect(Collectors.toList());
 
-    private final BitbucketClient client;
+    private ThreadLocal<BitbucketClient> client = new ThreadLocal<>();
```
The `ThreadLocal` seems unnecessary given there's no real need to hold the client state at class level. I'll make code suggestions for each change I think needs made to allow a the client to be defined at a method scope and passed around.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -74,10 +73,10 @@
 
-    public BitbucketServerPullRequestDecorator(BitbucketClient client) {
-        this.client = client;
+    public BitbucketPullRequestDecorator() {
```
Unnecessary empty constructor? Possibly delete this?

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -87,27 +86,44 @@ public String name() {
+
+        client.set(BitbucketClientFactory.createClient(bitbucketConfiguration));
+
         try {
-            if(!client.supportsCodeInsights()) {
-                LOGGER.warn("Your Bitbucket instances does not support the Code Insights API.");
+            if (!client.get().supportsCodeInsights()) {
+                LOGGER.warn("Your Bitbucket instance does not support the Code Insights API.");
+                client.remove();
```
Client simplification:
```suggestion
        BitbucketClient client = BitbucketClientFactory.createClient(bitbucketConfiguration);

        try {
            if (!client.suportsCodeInsights()) {
                LOGGER.warn("Your Bitbucket instance does not support the Code Insights API.");
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -87,27 +86,44 @@ public String name() {
-            client.createReport(project, repo,
-                    analysisDetails.getCommitSha(),
-                    toReport(analysisDetails)
+            CodeInsightsReport codeInsightsReport = client.get().createCodeInsightsReport(
+                    toReport(analysisDetails),
```
Client simplification:
```suggestion
            CodeInsightsReport codeInsightsReport = client.createCodeInsightsReport(
                    toReport(client, analysisDetails),
````

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -87,27 +86,44 @@ public String name() {
+                    analysisDetails.getQualityGateStatus()
             );
+
+            client.get().uploadReport(project, repo,
```
Client simplification:
```suggestion
            client.uploadReport(project, repo,
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -87,27 +86,44 @@ public String name() {
+            client.get().uploadReport(project, repo,
+                    analysisDetails.getCommitSha(), codeInsightsReport);
+
             updateAnnotations(project, repo, analysisDetails);
```
Client simplification:
```suggestion
            updateAnnotations(client, project, repo, analysisDetails);
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -87,27 +86,44 @@ public String name() {
             updateAnnotations(project, repo, analysisDetails);
         } catch (IOException e) {
             LOGGER.error("Could not decorate pull request for project {}", analysisDetails.getAnalysisProjectKey(), e);
+            client.remove();
```
Client simplification:
```suggestion
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -87,27 +86,44 @@ public String name() {
+            client.remove();
         }
 
+        client.remove();
```
Client simplification:
```suggestion
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -87,27 +86,44 @@ public String name() {
     }
 
-    private CreateReportRequest toReport(AnalysisDetails analysisDetails) {
+    private List<ReportData> toReport(AnalysisDetails analysisDetails) {
```
Client simplification:
```suggestion
    private List<ReportData> toReport(BitbucketClient client, AnalysisDetails analysisDetails) {
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -116,31 +132,24 @@ private CreateReportRequest toReport(AnalysisDetails analysisDetails) {
-                analysisDetails.getDashboardUrl(),
-                format("%s/common/icon.png", analysisDetails.getBaseImageUrl()),
-                asInsightStatus(analysisDetails.getQualityGateStatus()));
+        reportData.add(new ReportData("Analysis details", client.get().createLinkDataValue(analysisDetails.getDashboardUrl())));
```
Client simplification:
```suggestion
        reportData.add(new ReportData("Analysis details", client.createLinkDataValue(analysisDetails.getDashboardUrl())));
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -116,31 +132,24 @@ private CreateReportRequest toReport(AnalysisDetails analysisDetails) {
+        return reportData;
     }
 
     private void updateAnnotations(String project, String repo, AnalysisDetails analysisDetails) throws IOException {
```
Client simplification:
```suggestion
    private void updateAnnotations(BitbucketClient client, String project, String repo, AnalysisDetails analysisDetails) throws IOException {
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -116,31 +132,24 @@ private CreateReportRequest toReport(AnalysisDetails analysisDetails) {
         final AtomicInteger chunkCounter = new AtomicInteger(0);
 
-        client.deleteAnnotations(project, repo, analysisDetails.getCommitSha());
+        client.get().deleteAnnotations(project, repo, analysisDetails.getCommitSha());
```
Client simplification:
```suggestion
        client.deleteAnnotations(project, repo, analysisDetails.getCommitSha());
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -116,31 +132,24 @@ private CreateReportRequest toReport(AnalysisDetails analysisDetails) {
                 .map(componentIssue -> {
                     String path = componentIssue.getComponent().getReportAttributes().getScmPath().get();
-                    return new Annotation(componentIssue.getIssue().key(),
+                    return client.get().createCodeInsightsAnnotation(componentIssue.getIssue().key(),
```
Client simplification:
```suggestion
                    return client.createCodeInsightsAnnotation(componentIssue.getIssue().key(),
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecorator.java:
```diff
@@ -149,9 +158,9 @@ private void updateAnnotations(String project, String repo, AnalysisDetails anal
+        for (Set<CodeInsightsAnnotation> annotations : annotationChunks.values()) {
             try {
-                client.createAnnotations(project, repo, analysisDetails.getCommitSha(), new CreateAnnotationsRequest(annotations));
+                client.get().uploadAnnotations(project, repo, analysisDetails.getCommitSha(), annotations);
```
Client simplification:
```suggestion
                client.uploadAnnotations(project, repo, analysisDetails.getCommitSha(), annotations);
```

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketClient.java:
```diff
@@ -1,170 +1,82 @@
- * along with this program; if not, write to the Free Software Foundation,
- * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
- *
- */
```
This needs a copyright notice on it - please add one with your name and the current year

marvin-w replied:
Oups, strange. Will add one.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketClient.java:
```diff
@@ -1,170 +1,82 @@
+
+public interface BitbucketClient {
+
+    String REPORT_KEY = "com.github.mc1arke.sonarqube";
+    MediaType APPLICATION_JSON_MEDIA_TYPE = MediaType.get("application/json");
+    String TITLE = "SonarQube";
+    String REPORTER = "SonarQube";
+    String LINK_TEXT = "Go to SonarQube";
```
These don't feel like they should belong here as they're part of an implementation rather than an interface. I think Sonarqube may give you a warning about this too.

marvin-w replied:
The purpose of having them there was to use the same strings for both implementations: Cloud and Server. Since it's more or less the same functionality. I'd just duplicate them in both classes which is what I was trying to avoid. If sonar complains I'll change it - but I'll need to do some testing before that.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/model/BitbucketConfiguration.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    public boolean isCloud() {
+        return url.contains("https://api.bitbucket.org");
```
Should this be `startsWith` rather than `contains`? Given URLs are also case insensitive, I'd also suggest adding a `toLowerCase(Locale.ENGLISH)` in there.

marvin-w commented:
@mc1arke Thanks for the review again! I integrated your suggestions and fixed/adjusted the tests accordingly. Give it a go when you got some time :).

Please note that you need to set CloudCreateReportRequest.link to some valid link other than localhost for the API to fully function.

Otherwise you get below error:

`sonarqube    | com.github.mc1arke.sonarqube.plugin.ce.pullrequest.bitbucket.client.BitbucketCloudException: HTTP Status Code: 400; Message:{"key": "report-service.general.bad-request", "message": "link is not a valid URL", "arguments": {}}
`

yeroc commented:
@marvin-w Would love to try this out.  It looks like we'll need to be running SonarQube 7.8 or newer to try this out?  We're running 7.7 right now so will need to upgrade first I guess?

marvin-w commented:
Yes, you'll need to upgrade I believe. I didn't test it in lower versions.

marvin-w commented:
@mc1arke would you mind retriggering the sonar analysis? For some reason it doesn't analyse the code in this branch.

mc1arke commented on src/test/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketServerClientUnitTest.java:
```diff
@@ -0,0 +1,400 @@
+
+
+        // when
+        try {
+            underTest.uploadReport("project", "repository", "commit", report);
+            fail();
+        } catch (BitbucketException e) {
+            assertTrue(e.isError(400));
+            assertEquals("error!", e.getMessage());
+        }
```
Potentially use AssertJ here:

```suggestion
        assertThatThrownBy(() -> underTest.uploadReport("project", "repository", "commit", report))
                .isInstanceOf(BitbucketException.class)
                .hasMessage("error!")
                .extracting(e -> ((BitbucketException) e).isError(400))
                .isEqualTo(true);
```
I'm also not sure what the different between this and the previous test method is, but a similar change could be made there,

marvin-w replied:
That looks way cleaner thank you!

mc1arke commented on src/test/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketServerClientUnitTest.java:
```diff
@@ -0,0 +1,400 @@
+    }
+
+    @Test
+    public void testCreateCloudReport() throws IOException {
```
`IOException` is never thrown by this method?

mc1arke commented on src/test/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketServerClientUnitTest.java:
```diff
@@ -0,0 +1,400 @@
+    }
+
+    @Test
+    public void testCreateDataLinkForServer() throws IOException {
```
`IOException` is never thrown by this method?

marvin-w commented:
@mc1arke Thank you for review. I took care of all the points you mentioned. Would you mind retriggering the sonar analysis? I added missing tests for the _BitbucketCloudException_ and the _BitbucketClientFactory_.

Also, what is your opinion concerning the documentation mentioned in the description of this pull request? Since this feature is not officially supported by SonarQube (i.e. no official docs) it might become necessary.

I rebased all the commits into one and did a final test after the refactoring to be sure that the functionality is still working as intended.

marvin-w commented:
@mc1arke You're probably a busy man and I'm sorry for bothering you again but do you have an ETA on when we can expect this to be merged and released?

ajcamilo commented:
I've tested your PR with Sonarqube 7.9.3 with Bitbucket Cloud and it seems to work. 

Except for the annotations. I don't know if it's something I need to change in bitbucket, but the annotations are not shown in the comments.

marvin-w commented:
Cool, glad it works for you. There are no comments on files. The annotations belong to the report. In the report you can go directly to the specific file and line.

Please look at my screens in the PR description. If you still think there is an issue please add some more information to it, maybe a screenshot.

ajcamilo commented:
You are right, sorry. I was looking at screenshots of the Bitbucket Server (adds annotations to the diff section)

Everything worked :+1:  

Maybe there should be some information for the format the bitbucket token (base64 user:pass), because this information will be lost in this PR.

marvin-w commented:
Exactly that is what I meant in https://github.com/mc1arke/sonarqube-community-branch-plugin/pull/185#issuecomment-643642458.

mc1arke commented:
Merged. Many thanks for the contribution!

Now that you're a contributor to this repository, I believe you should have access to edit the Wiki. If you don't then please let me know and I'll look at options that allow easier contribution to documentation.

marvin-w commented:
@mc1arke Thanks for taking care of it. I have two more things to mention here.

1. I am able to pull the wiki but I can't edit it, neither in the UI nor in my local git clone. When forking the repo the wiki is also _not_ included and I wouldn't be able to create a pull request. I could still pull your wiki and force push it to my fork, do the changes but you'll then need to do the merging manually. For me, personally, this would be a pain so I believe that there should be something else. I've used custom wikis deployed via https://www.netlify.com/ before with great success. If you're interested in that I could create you an initial PR and you could have a look. This would have the benefit that everyone can contribute to the documentation and you also have a live preview of all the changes in any given feature branch.

2. Should I create a PR for the merge from release/1.3 to master in order to enable support for bitbucket cloud PR decoration also for Sonarqube >7.9 or do you want to take care of that yourself?

yeroc commented:
@marvin-w @mc1arke Thanks so much for your work on it, I've built and tested this on a Sonar 7.9.3 (LTS) installation and it works great.  Look forward to seeing this in a release to simplify deployment!

raivishnu commented:
Hi @marvin-w / @mc1arke   I am trying to use it for PR Decoration for gradle based Java Project. I have pulled the latest code build and deployed in the server and can it is properly showing the branches. I need help in setting it up for PR Decoration. The current Wiki does not have clear instruction to set this up. I will appreciate your help in this regard.

- Bitbucket Cloud (Private Repository) 

- Sonarqube Server V 8.3.1 

- Jenkins (Pipleine)

anuragpathak21 commented:
Hi @marvin-w / @mc1arke   I am trying to use it for PR Decoration for gradle based Java Project. I have pulled the latest code build and deployed in the server and can it is properly showing the branches. I need help in setting it up for PR Decoration. The current Wiki does not have clear instruction to set this up. I will appreciate your help in this regard.

- Bitbucket Cloud (Private Repository) 

- Sonarqube Server V 8.3.1 

- Jenkins (Pipleine)

marvin-w commented:
@anuragpathak21 This feature is currently only supported in the LTS 7.9.x version.

anuragpathak21 commented:
> @anuragpathak21 This feature is currently only supported in the LTS 7.9.x version.

I am planning to downgrade to LTS 7.9. Will appreciate if you provide details of the setup.

marvin-w commented:
@anuragpathak21 Give me a day I'll come back to you.

marvin-w commented:
For the general setup of the PR decoration please refer to the official SonarQube documentation.

For the specific feature you need a user with the "Read Repository" Scope. The token is defined as `base64(username:password)` of your technical user. I also wrote that https://github.com/mc1arke/sonarqube-community-branch-plugin/pull/185#issuecomment-640257544 before (in this very thread).

For anything else please refer to the respective tool documentation, i.e. if there's something wrong in Jenkins ask them or if you also believe there should be PR refs in bitbucket cloud then this is the right way for you: https://jira.atlassian.com/browse/BCLOUD-5814. Without this it is very hard to integrate the PR decoration with Jenkins as there are no dedicated git refs for pull requests.

myDisconnect commented:
@yeroc @marvin-w how did you manage to build this under 7.9.3 (LTS)? 
I tried changing the `build.gradle` ->  `sonarqubeVersion` to 7.9.3 and got so many exceptions.

yeroc commented:
@myDisconnect I suspect you're compiling the wrong branch.  You need to clone/checkout and build the `release/1_3` branch.

marvin-w commented:
@mc1arke When can we expect this to be released? Is there anything we can help with?

mc1arke commented:
@marvin-w This has now been released in 1.3.2 (Sonarqube 7.8-8.0) and 1.4.0 (Sonarqube 8.1). Sorry for the delay with getting the releases out, and thanks once again for the contribution.

On the documentation side: I'd like the documentation to be tightly linked to a release version of possible, so that it's clear how to configure a certain version of a plugin by looking at the relevant tag in Git. I'd therefore envisaged having a `docs` directory (or similar) in Git containing Markdown file. Does this cause any particular issues for you?

marvin-w commented:
@mc1arke Thanks! No worries :)

I'll also have a look concerning the documentation then.

rhaex commented:
> @mc1arke I'm done with this. Feel free to review it once you have some time.
> 
> @singhsoldier13, @goober, @jburgel-davinci, @yeroc If there is a chance it would be awesome if you could also go for an additional test run with this version of bitbucket cloud PR decoration.
> 
> For the token please use `base64(username:password)` (At least Read repository scope is required)
> 
> For the URL please use `https://api.bitbucket.org`

Could this information please be added to the readme.MD file? As it points to the official documentation, but the official version does not support BitBucket cloud, so the only place where to find these parameters right now is this pull-request.

nfalco79 commented:
@marvin-w if this PR also approve the Pull request (configurable) without any kind of comment was great. Because Repose code insight does not play in any way in branch restriction.

marvin-w commented:
> @marvin-w if this PR also approve the Pull request (configurable) without any kind of comment was great. Because Repose code insight does not play in any way in branch restriction.

I read your post three times and I don't get what you are trying to tell your audience in a PR that has been closed for over a year.

nfalco79 commented:
Testo di origine
I would say that approval should not be discarded as idea. Ok that bitbuket "Code insights" is the tool that bitbuket provides to report any kind of metric but it does not play any role in branch restrictions. This means for example that you could have a PR green just because the code compile, regardless if Sonarqube reports failures or not. PR could be merged and there are not option to prevent this. The approval instead could play a role in a branch restrictions. I mean the approval code could be left there (configurable) based on quality gate.

marvin-w commented:
Please open an issue for additional feature requests or bug reports.

nfalco79 commented:
We have migrated the sonarqube version for a month from the mibexsoftware plugin to this one and I must say that most developers are still puzzled to understand when there are actually analisy sonars. Both because the Pull Request Decoration must be configured in each project (and this is a problem because it must be done after the project has been created) to see the reports and because it is not very visible in BB Cloud.
We then had to implement the sonar analysis pipeline stage decoration as failed (but not the build) and then implement the PR approval by waiting for the analysis id to be available to reach the minimum 2 approvals for a PR, one automatic (build system) and the other of another real person. If it had been possible to enable approval directly in the plugin it would have been easier since the plugin already has all the necessary information. 

> Please open an issue for additional feature requests or bug reports.

Sure I will do

nfalco79 commented:
Issue #77 is still open and summarize above requirement
